packages/peer-exchange/src/waku_peer_exchange.ts

```diff
@@ -0,0 +1,133 @@
+import type { Stream } from "@libp2p/interface-connection";
+import { PeerId } from "@libp2p/interface-peer-id";
+import type { Peer, PeerStore } from "@libp2p/interface-peer-store";
+import type { IncomingStreamData } from "@libp2p/interface-registrar";
+import {
+  getPeersForProtocol,
+  selectConnection,
+  selectPeerForProtocol,
+} from "@waku/core";
+import { ENR } from "@waku/enr";
+import {
+  PeerExchange,
+  PeerExchangeComponents,
+  PeerExchangeQueryParams,
+  PeerExchangeResponse,
+  ProtocolOptions,
+} from "@waku/interfaces";
+import debug from "debug";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { PeerExchangeRPC } from "./rpc.js";
+
+export const PeerExchangeCodec = "/vac/waku/peer-exchange/2.0.0-alpha1";
+
+const log = debug("waku:peer-exchange");
+
+export class WakuPeerExchange implements PeerExchange {
+  private callback:
+    | ((response: PeerExchangeResponse) => Promise<void>)
+    | undefined;
+
+  constructor(
+    public components: PeerExchangeComponents,
+    public createOptions?: ProtocolOptions
+  ) {
+    this.components.registrar
+      .handle(PeerExchangeCodec, this.onRequest.bind(this))
+      .catch((e) => log("Failed to register peer exchange protocol", e));
+  }
+
+  async query(
+    params: PeerExchangeQueryParams,
+    callback: (response: PeerExchangeResponse) => Promise<void>
+  ): Promise<void> {
+    const { numPeers } = params;
+
+    const rpcQuery = PeerExchangeRPC.createRequest({ numPeers });
+
+    const peer = await this.getPeer();
+
+    const stream = await this.newStream(peer);
+
+    await pipe(
+      [rpcQuery.encode()],
+      lp.encode(),
+      stream,
+      lp.decode(),
+      async (source) => await all(source)
+    );
+
+    this.callback = callback;
+  }
+
+  private onRequest(streamData: IncomingStreamData): void {
```

Comment:
```suggestion
  private handler(streamData: IncomingStreamData): void {
```

Not sure `onRequest` is a great name (I just got confused by it)